src/index.js

```diff
@@ -14,73 +14,25 @@ const map = require('async/map')
 const series = require('async/series')
 const waterfall = require('async/waterfall')
 const MemoryStore = require('interface-datastore').MemoryDatastore
+const mergeOptions = require('merge-options')
+const ipldDagCbor = require('ipld-dag-cbor')
+const ipldDagPb = require('ipld-dag-pb')
 
 function noop () {}
 
 class IPLDResolver {
-  constructor (options) {
+  constructor (userOptions) {
+    const options = mergeOptions(IPLDResolver.defaultOptions, userOptions)
```

Comment:
This should be enough here right ?
```suggestion
    const options = Object.assign({}, IPLDResolver.defaultOptions, userOptions)
```

Reply:
I want a to have a deep copy. Example:

```js
'use strict'

const mergeOptions = require('merge-options')

const oa1 = {main: {sub: 1}}
const oa2 = {main: {sub: 2}}
const oa = Object.assign({}, oa1, oa2)
oa2.main.sub = 3
console.log('oa:', oa)

const mo1 = {main: {sub: 1}}
const mo2 = {main: {sub: 2}}
const mo = mergeOptions(mo1, mo2)
mo2.main.sub = 3
console.log('mo:',  mo)
```

Output is:

```js
oa: { main: { sub: 3 } }
mo: { main: { sub: 2 } }
```

See https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Object/assign#Deep_Clone for more information about `Object.assign()`.

Reply:
ok just wanted to make sure we really need deep copy here.

Reply:
K, cool. I prefer code which doesn't lead to weird surprises in the long run (+ I've coded functional programming language for a looong tim)

Reply:
@vmx can i suggest https://www.npmjs.com/package/assign-deep instead of merge-options. Check https://codesandbox.io/s/742p0rjvy0, assign-deep is smaller and handles non-plain-objects values better like promises (merge-options gives empty object when you should get a Promise). We should use assign-deep everywhere.

Reply:
@hugomrdias: I want to have a deep copy, that makes things easier to debug/reason about. `assign-deep` does only a shallow copy.

Reply:
I just read https://github.com/ipfs/js-ipfs/pull/1663#discussion_r228094561. Perhaps it's not always a good idea to copy things. I think I need to put more thought into this. Though I'll merge this for now to keep things going.